[16.0][IMP] shopfloor/mobile: Allow to do single unload with each pack

rousseldenis opened the pull request.

rousseldenis committed:
[IMP] shopfloor_mobile: Improve package detail card presentation

Add a package icon
Allow to change color of card

rousseldenis committed:
[IMP] shopfloor_mobile: Improve package detail card presentation

Add a package icon
Allow to change color of card

rousseldenis committed:
[IMP] shopfloor/mobile: Allow to do single unload with each pack

rousseldenis committed:
[IMP] shopfloor/mobile: Allow to do single unload with each pack

jbaudoux commented on shopfloor/models/shopfloor_menu.py:
```diff
@@ -134,6 +140,12 @@ class ShopfloorMenu(models.Model):
         default=False,
         help=UNLOAD_PACK_AT_DEST_HELP,
     )
+    unload_single_package_is_possible = fields.Boolean(
+        compute="_compute_unload_single_package_is_possible"
+    )
+    unload_single_package = fields.Boolean(
+        string="Unload single package", help=UNLOAD_SINGLE_PACKAGE_HELP
+    )
```
you need to declare those fields in the data.xml for scenario_cluster_picking

rousseldenis replied:
@jbaudoux Done

rousseldenis committed:
[IMP] shopfloor: In cluster pickings, allows to select another package at unload

To be more flexible as the user has a particular package in front of him,
he can scan another package than the one proposed.

rousseldenis committed:
[IMP] shopfloor: In cluster pickings, allows to select another package at unload

To be more flexible as the user has a particular package in front of him,
he can scan another package than the one proposed.

rousseldenis committed:
[IMP] shopfloor: Add configuration data for unload single package options

rousseldenis committed:
[IMP] shopfloor: Add configuration data for unload single package options

jbaudoux commented on shopfloor_mobile/static/wms/src/i18n/fr.json:
```diff
@@ -14,6 +14,7 @@
   "cluster_picking.unload_all.title": "Décharger tous les compartiments",
   "cluster_picking.unload_set_destination.title": "Définir un emplacement de destination",
   "cluster_picking.unload_single.title": "Décharger un seul compartiment",
+  "cluster_picking.unload_scan_pack.scan_placeholder": "Scannez la boîte",
```
```suggestion
  "cluster_picking.unload_scan_pack.scan_placeholder": "Scannez le compartiment",
```

jbaudoux commented on shopfloor_mobile/static/wms/src/i18n/en.json:
```diff
@@ -16,6 +16,7 @@
   "cluster_picking.unload_all.title": "Unload all bins",
   "cluster_picking.unload_set_destination.title": "Set destination",
   "cluster_picking.unload_single.title": "Unload single bin",
+  "cluster_picking.unload_scan_pack.scan_placeholder": "Scan Package",
```
```suggestion
  "cluster_picking.unload_scan_pack.scan_placeholder": "Scan bin",
```

jbaudoux reviewed (commented):
i think you should differentiate:
* `unload_scan_bin` base shopfloor where you have put products in bins
* `unload_scan_package` in case you repack at the end of the picking

rousseldenis committed:
[IMP] shopfloor: Cluster Picking : Get the selected move line to allow retrieving more information

rousseldenis committed:
[IMP] shopfloor: Cluster Picking : Get the selected move line to allow retrieving more information

rousseldenis committed:
[IMP] shopfloor_mobile: Cluster Picking: Add a enriched card for destination location

In order to enrich the presentation of the destination location, replace the location
name with a component

rousseldenis committed:
[IMP] shopfloor_mobile: Cluster Picking: Add a enriched card for destination location

In order to enrich the presentation of the destination location, replace the location
name with a component

nicolas-delbovier-acsone committed:
[IMP] shopfloor_mobile: add missing color to <detail-location-destination-operation> card

nicolas-delbovier-acsone committed:
[IMP] shopfloor_mobile: add missing color to <detail-location-destination-operation> card

nicolas-delbovier-acsone committed:
[FIX] shopfloor_mobile_base: display falsy but not undefined fields in detail-card.

rousseldenis committed:
[FIX] shopfloor: Don't let package be unloaded twice.

If user try to scan the package just having been unloaded,
the batch will be marked as done as no more lines would be found.

Don't let the user to scan it one more time.

rousseldenis committed:
[FIX] shopfloor: Don't let package be unloaded twice.

If user try to scan the package just having been unloaded,
the batch will be marked as done as no more lines would be found.

Don't let the user to scan it one more time.

rousseldenis committed:
[FIX] shopfloor: In cluster picking, don't authorize package unload twice

If the last unload validated the concerned picking, the reuse of package
in that one will conduct to batch validation.

rousseldenis committed:
[FIX] shopfloor: In cluster picking, don't authorize package unload twice

If the last unload validated the concerned picking, the reuse of package
in that one will conduct to batch validation.

rousseldenis committed:
[FIX] shopfloor: In cluster picking, don't authorize unload scan destination twice

rousseldenis committed:
[FIX] shopfloor: In cluster picking, don't authorize unload scan destination twice

lmignon reviewed (approved):
LGTM (Code + Functional)

simahawk commented on shopfloor_mobile/static/wms/src/components/detail/detail_package.js:
```diff
@@ -8,6 +8,9 @@ import {ItemDetailMixin} from "/shopfloor_mobile_base/static/wms/src/components/
 Vue.component("detail-package", {
     mixins: [ItemDetailMixin],
+    props: {
+        color: String, // Used to determine card_color (See `utils.colors.color_for()`)
```
Any reason not to add `detail_main_card` as default? It will change existing screens.
However, we should probably put the default in `item-detail-card`

simahawk reviewed (commented):
LG overall, just code review.
Small remark on the card color default.

jbaudoux commented on shopfloor/models/stock_picking_batch.py:
```diff
@@ -39,3 +40,14 @@ def _compute_picking_info(self):
+        Returns the first package in batch lines from the barcode value
+        """
+        self.ensure_one()
+        return first(
+            self.move_line_ids.result_package_id.filtered(
+                lambda package: package.name == barcode
+            )
+        )
```
Would be nice to have `for_packages` in https://github.com/OCA/wms/pull/1174 and let `package_from_scan` resolve the barcode. For GS1 compliancy.

cc @nicolas-delbovier-acsone

jbaudoux commented on shopfloor_mobile/static/wms/src/scenario/cluster_picking.js:
```diff
@@ -383,7 +423,7 @@ const ClusterPicking = {
                         title: this.$t("cluster_picking.unload_single.title"),
                         scan_placeholder: this.$t(
-                            "cluster_picking.scan_destination.scan_placeholder"
+                            "cluster_picking.unload_scan_pack.scan_placeholder"
```
Could you solve this:
* unload_scan_bin : base shopfloor where you have put products in bins
* unload_scan_package : in case you repack at the end of the picking
